### DISCUSSION TITLE
ESP32 async SSL (as server) and PC SSL via Wifi (as client)

### DISCUSSION THREAD

**Discussion by nastiliano** [micropython/micropython#13480] [General]
Hello,

After a long time trying to connect my ESP32 device as a client to a MQTT broker with TLS (self-signed certs), finally (thanks lot of people in this forums) I could connect and work: https://github.com/orgs/micropython/discussions/13441

Now I wondering if it's possible too use the same certificates but in annother situation. I wanted to send some information via Wifi from a  PC (acting as client) using python "paho" to my ESP32 device that is configured as a server in a async sockets situation.

Yesterday I tried a lot, searching info on the net, and I think I was near to do it, but I think I need some more specific configuration about type of cypher (EC) etc...

Is possible sending encrypted data by Wifi?

Thank you very much

**Comment by nastiliano** [comment]
Hello,

Sorry, @Carglglz could just give me some link, please?I saw you are so experienced in that.

Thank you very much

**Reply by Carglglz** [reply]
> Now I wondering if it's possible too use the same certificates but in annother situation. I wanted to send some information via Wifi from a PC (acting as client) using python "paho" to my ESP32 device that is configured as a server in a async sockets situation.

You definitely can, isn't `paho` an MQTT client library?, just connect the PC client to the same MQTT broker and publish to a topic that the ESP32 is subscribed. 

> Is possible sending encrypted data by Wifi? 

Yes that is what  MQTT broker with TLS does. 

If you don't want to use MQTT but instead a direct connection between PC client and the ESP32 device then just use [microdot](https://github.com/miguelgrinberg/microdot) as a Websocket server and a Websocket client from [aiohttp](https://docs.aiohttp.org/en/stable/) for Python 3 and here for MicroPython [micropython-lib/aiohttp](https://github.com/micropython/micropython-lib/tree/master/python-ecosys/aiohttp)

https://microdot.readthedocs.io/en/latest/extensions.html#websocket-support
https://docs.aiohttp.org/en/stable/client_quickstart.html#websockets

**Reply by nastiliano** [reply]
Hello,

Thanks in advance for fast respond. I apreciate.

1st) Yes of course, I'm already connecting PC and ESP32 via MQTT (TLS) is perfect. Thank you very much.

2nd) Ok, thanks, but I wanted to connect not with a websocket server, I wanted to create a connection directly by Wifi with TLS encrypted data. The ESP32 device as a async socket server and PC as a client. Now I'm doing this by my own encryption simply using RSA encryption, then Python sockets sending to my ESP32 as a async server, but I wondered if I could do something more sofisticaded wraping sockets in TLS protocol.

Thanks, best regards.

**Reply by Carglglz** [reply]
> but I wondered if I could do something more sofisticaded wraping sockets in TLS protocol.

Yes of course then just use `SSLContext` see [ssl](https://docs.micropython.org/en/latest/library/ssl.html)  and e.g. [asyncio_tls_server_client.py](https://github.com/micropython/micropython/blob/master/tests/multi_net/asyncio_tls_server_client.py) 👍🏼

**Reply by nastiliano** [reply]
Hello again :)

After dozens of tries (changing little thigs), I've got always the same error in my ESP32 device (as a server). It seems that async coro "handle_connection" tries to receive but there's a problem with TLS certificates and the memory of device is overflows (at least this is that I believe, heheh...):

    Traceback (most recent call last):
        File "asyncio/stream.py", line 1, in _serve
    OSError: [Errno 12] ENOMEM

My PC code (client):

    # Create an SSL context
    context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    context.verify_mode = ssl.CERT_REQUIRED
    context.check_hostname = False

    # Load CA certificate with which the client will validate the server certificate
    context.load_verify_locations(os.path.join(folder_certs  , "ec_cert.crt"))

    # Create a client socket
    self.socket = socket.socket (socket.AF_INET, socket.SOCK_STREAM)

    # Make the client socket suitable for secure communication
    self.socket  = context.wrap_socket(self.socket)
    
    # Data
    self.data = "hello world"

    with self.socket:
        # Timeout
        self.sosket.settimeout(5)

            try:
            self.socket((self.IP , 8000))
            self.socket(self.data)

My ESP32 code (server):

    async def main():

       # wifi config
       # ....

        async def handle_connection(reader, writer):
            data = await reader.read(100)
            print("echo:", data)

        cert = cafile = "/certs/ec_cert.der"
        key = "/certs/ec_key.der"

        os.stat(cafile)
        os.stat(key)

        server_ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
        server_ctx.load_cert_chain(cert, key)

        server = await asyncio.start_server(handle_connection, "192.168.1.200", 8000 , ssl = server_ctx)
        await asyncio.gather(server , otherts())

    asyncio.run(main())


Any clue please?  :)

Thank you very much.

**Reply by nastiliano** [reply]
Hello,

Just the last little help please @Carglglz?

Thank you very much.
